snapcraft/internal/meta.py

```diff
@@ -207,6 +211,33 @@ def _wrap_exe(self, command, basename=None):
 
         return os.path.relpath(wrappath, self._snap_dir)
 
+    def _generate_hook(self, hook, command):
+        hooks_dir = os.path.join(self._snap_dir, 'meta', 'hooks')
+        os.makedirs(hooks_dir, exist_ok=True)
+
+        hook_path = os.path.join(hooks_dir, hook)
+        with contextlib.suppress(FileNotFoundError):
+            os.remove(hook_path)
+
+        command_parts = shlex.split(command)
+
+        shebang = None
+        hook_exec = '$SNAP/{}'.format(command_parts[0])
+        exepath = os.path.join(self._snap_dir, command_parts[0])
+        if not os.path.exists(exepath) and '/' not in command_parts[0]:
+            _find_bin(command_parts[0], self._snap_dir)
+            hook_exec = command_parts[0]
+        else:
+            with open(exepath, 'rb') as f:
+                # If the file has a she-bang, the path might be pointing to
+                # the local 'parts' dir. Extract it so that _write_wrap_exe
+                # will have a chance to rewrite it.
+                if f.read(2) == b'#!':
+                    shebang = f.readline().strip().decode('utf-8')
```

Comment:
can we refactor these lines into a function so they are not duplicated in wrap_exe?